### ISSUE TITLE
Make HttpClient use a singleton SocketsHttpHandler

### ISSUE THREAD

**Issue by scalablecory** [dotnet/runtime#33124]
## Proposal

We update the parameterless `HttpClient` constructor to use a singleton `SocketsHttpHandler`.

## Why

There are well known articles online discussing how to use `HttpClient` correctly: specifically, recommending to keep a singleton/etc. around instead of using a per-request client. It is very common on Stack Overflow for people to comment about this when a question involving `HttpClient` comes up. `HttpClient` has effectively become a [missing stair](https://en.wikipedia.org/wiki/Missing_stair).

This change would make the API behave closer to peoples expectations: new users would not step on this landmine, and we may even fix some performance problems in existing code. It brings us in line with how people use e.g. `SqlConnection` and Entity Framework today, which use global pools very effectively.

## Why not

This will be a subtle change of behavior for anyone using `HttpClient.Dispose()` to explicitly manage resource usage: instead of clearing resources immediately, it would keep them alive for some minutes until idle connection lifetime limits cleans them up.

Users looking for explicit fresh DNS lookups would be unable to use this.

We also know that some Azure environments have a limit on the number of outbound connections that can be made against a single IP, and these advanced users may be depending on careful management of pooling.

We might mitigate this by having an `AppContext` switch, but ultimately there are some (presumably small, but important) numbers of people who will try a "free lunch" upgrade and hit this as a problem.

## Alternatives

- Analyzers might be useful here, but it would be difficult to avoid false positives and so such an analyzer might end up off by default and thus not helpful to those users who would need it most. 
- We could add static methods onto `HttpClient`. I suspect most users don't need to set `HttpClient` options, so this would get usability 90% of the way there. This would not help existing code, and would probably miss a lot of new users who're reading how to use `HttpClient` from existing material.

CC @dotnet/ncl @stephentoub opinions?

**Comment by wfurt** [comment]
_If_ DNS update is one of the cons I'm wondering that is something we can also improve. For example, do DNS lookup more frequently and not open a new connection when the name resolves to the IP of an existing pool connection. Of course, getting API when we can get TTL for DNS would be better but we may be able to make some improvements without it.

**Comment by stephentoub** [comment]
> We update the parameterless HttpClient constructor to use a singleton SocketsHttpHandler.

You mean HttpClientHandler (which on some platforms wraps SocketsHttpHandler), right?

**Comment by scalablecory** [comment]
> > We update the parameterless HttpClient constructor to use a singleton SocketsHttpHandler.
> 
> You mean HttpClientHandler (which on some platforms wraps SocketsHttpHandler), right?

Yes. I guess we need to consider the effects of this from Xamarin perspective as well.

**Comment by geoffkizer** [comment]
Interesting suggestion. I think the interesting question is, how often do users use the parameterless HttpClient constructor and implicit handler, vs using their own handler? I suspect we could get API usage data on this.

This does introduce a "cliff" where as soon as you switch to creating the handler yourself, you'll no longer get the benefits of this, and that might be surprising to some users. But also, it seems reasonable to say that users who create their own handler probably also understand how to properly manage HttpClient lifetime.

**Comment by kjpou1** [comment]
> > > We update the parameterless HttpClient constructor to use a singleton SocketsHttpHandler.
> > 
> > 
> > You mean HttpClientHandler (which on some platforms wraps SocketsHttpHandler), right?
> 
> Yes. I guess we need to consider the effects of this from Xamarin perspective as well.

Also, keep in mind that the WebAssembly implementation does not use the SocketsHttpHandler.

**Comment by Timovzl** [comment]
Perhaps there is an alternative that would handle the DNS issue as well. What I am about to suggest is intended to make naively-constructed `HttpClient`s behave somewhat like ones produced by `HttpClientFactory`.

- Use the behavior described here when an `HttpClient` is constructed through the default constructor.
- _Avoid_ assigning a new `HttpMessageHandler` in the constructor.
- Have a (lazy) pool of `HttpMessageHandler`s, intended for just this scenario.
- The handlers in the pool have a limited time to live (TTL), after which they are removed and disposed.
- When an `HttpClient` makes a request, it takes from the pool a handler with TTL at least as great as the request timeout.
- The pool attempts to balance shorter TTLs (fresher DNS) with limiting the number of handlers in the pool.
- Rather than retaining the handler, the `HttpClient` uses it only for the request itself. (After all, when the next request is made, it must consider the `Timeout` value at that time, as well as the remaining TTLs of handlers in the pool.)

Clearly, using `HttpClientFactory` is superior. However, this gets us close to its benefits for existing code and naive instantiation, removing most of the booby traps.

**Comment by MihaZupan** [comment]
That is about the same as keeping a single `SocketsHttpHandler` instance with a reasonable `PooledConnectionLifetime` set.

**Comment by Timovzl** [comment]
@MihaZupan, now I feel silly going through all this trouble. 😛

A reasonable `PooledConnectionLifetime` on the proposed handler seems ideal.

Although, come to think of it, I'm not sure such a property is available on all message handler implementations, i.e. on all platforms...

**Comment by MihaZupan** [comment]
It's supported on all major platforms except Browser (client-side Blazor) and Mobile (not used by default), but those kind of applications shouldn't be running into issues like port exhaustion (primary reason to make such a change).
If there are issues on such platforms that this may impact, we would certainly want to account for that.

**Comment by wsq003** [comment]
Since this Proposal did not come true for over a year, maybe it is not appropriate.
It would be great if this proposal come true, after careful consideration.
If this proposal is not likely to come true, I would like to suggest #56284 as an alternative.

What really bad here is that programmer may unconsciously write some code with fatal error.
So I believe the most import thing is to warn user about such implicit fatal error.

**Comment by DL444** [comment]
Cookie handling might be a problem here. From my understanding automatic cookie handling is implemented by HttpClientHandler. If so, transparent handler pooling can break applications that expect the cookie container to be clean each time they new up an HttpClient.

Docs for IHttpClientFactory state that you either disable automatic cookie handling or avoid using the factory. The former would require non-trivial code changes to implement, so it wouldn't be helpful in this case.
